---
title: "Introduction to biscuit"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Installation

`biscuit` is an R package that utilizes a Bayesian hierarchical inference model to analyze high-throughput CRISPR screen data. It performs inference using Stan via the `cmdstanr` interface.

To install `cmdstanr` (and `cmdstan`):

```{r}
# we recommend running this is a fresh R session or restarting your current session
install.packages("cmdstanr", repos = c('https://stan-dev.r-universe.dev', getOption("repos")))

library(cmdstanr)
install_cmdstan(cores = 2) # number of cores to use for installation
```

To install `DESeq2`:

```{r}
 if (!require("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
 }

BiocManager::install("DESeq2")
library(DESeq2)
```

To install `biscuit`:

```{r}
if (!requireNamespace("remotes", quietly = TRUE)) {
  install.packages("remotes")
}

remotes::install_github("sandyskim/biscuit")
library(biscuit)
```

## Using `biscuit`

Here is the typical workflow when using `biscuit`:

1.  **Load and bundle data into a `dough` object**\
    Organize raw counts, guide-to-gene mapping, and sample design into a single structured object.

2.  **Filter and preprocess the `dough` object**\
    Remove low-count guides and genes with few guides.

3.  **Fit the `biscuit` model**\
    Apply the hierarchical Bayesian model to infer guide- and gene-level effects.

4.  **Explore posterior estimates**\
    After fitting the model, summarize the guide- and gene-level posterior distributions, which provide both effect size estimates and uncertainty, and significant hits. Additionally, `biscuit` provides plotting functions to visualize these results.

To demonstrate how to use `biscuit`, we provide an example using the data from Regnase-1 screen (<https://doi.org/10.1038/s41586-019-1821-z>).

### Input

First, load in the data from the Regnase-1 screen:

```{r}
data(regnase)
```

To create a `dough` object, you need three inputs:

-   **Raw sequencing counts**: a matrix or data frame with `n_guides` rows (guides) and `n_samples` columns (samples).\
-   **Guide-to-gene mapping**: a two-column matrix or data frame with `n_guides` rows, containing the guide name and its corresponding gene name.\
-   **Sample design**: a vector of length `n_samples` specifying the experimental condition for each sample (must include exactly two conditions, e.g., *control* and *treatment*).\
-   *(Optional)* **Non-targeting controls**: a list of guide names corresponding to non-targeting controls. If your screen includes non-targeting controls, we strongly recommend including them, as they improve inference.

To make a `dough` object, we call the function `make_dough()`:

```{r}
dough_obj <- make_dough(
  counts = regnase$counts,
  guides = regnase$guides,
  design = regnase$design,
  ntcs = regnase$ntcs
)
```

### Filtering data

Now that the data is organized in a `dough` object, we can filter the data before fitting the model.

`biscuit` performs the following preprocessing steps:

-   **Removing low-count guides**: guides with very few reads across all samples may not provide reliable information. By default, guides with \<30 counts are filtered out.
-   **Filtering genes with few guides**: genes targeted by only one or two guides can be noisy. By default, genes with \<2 guides targeting it are filtered out.

To preprocess the data, we call the function `trim_dough()`:

```{r}
dough_obj <- trim_dough(
  dough_obj,
  min_counts = 30, 
  min_guides_per_gene = 2, 
  verbose = TRUE
)
```

### Running `biscuit`

Once the `dough` object is prepared, it is ready to fit the model.

First, we specify the directory in which the samples and logs will be saved:

```{r}
output_dir <- getwd()
```

To fit the model, we call the function `bake_biscuit()`. `biscuit` uses a Bayesian hierarchical model to infer guide-level and gene-level effects from CRISPR screen data.

Note: `biscuit` performs statistical inference using MCMC sampling, which can be computationally intensive. Depending on your computational resources, fitting the model on this data set may take *up to an hour or more*.

```{r}
biscuit_fit <- bake_biscuit(
  dough_obj, 
  output_dir,
  save_samples=TRUE, 
  n_parallel_chains=4,
  seed=13, 
  pseudocount=TRUE)
```

### Exploring the posterior

Once the model is finished fitting, `biscuit` computes posterior summaries as well as the local false sign rate for each guide-level and gene-level effect for downstream analysis.

#### Summarizing posterior estimates

`biscuit` computes a summary table of the posterior for the modelâ€™s parameters of interest:

-   `mu`: gene-level effects
-   `beta1`: guide-level effects
-   `beta0`: baseline guide-level expression (intercept)
-   `phi`: guide-level dispersion
-   `gamma`: sample-level dispersion scaling factors

For each parameter, the summary table reports the **mean**, **median**, **standard deviation**, and **2.5% and 97.5% quantiles**. In addition, the **local false sign rate** is calculated for `beta1` and `mu`, providing a measure of significance with directional confidence.

To summarize posterior draws and calculate LFSR, we call the function `summarize_params()`:

```{r}
biscuit_fit <- summarize_params(biscuit_fit, pars = c("mu", "beta1", "beta0", "phi", "gamma"))

# view the summary of beta1 (guide-level effects)
head(biscuit_fit$results$beta1)

# view the summary of mu (gene-level effects)
head(biscuit_fit$results$mu)
```

#### Plotting results

After `biscuit` summarizes the posterior draws for the parameters of interest, `biscuit` provides a few plotting functions to visualize results:

`plot_guide_violin()` shows the distribution of the posterior mean effects for each guide (`beta1`), if applicable, stratified by targeting and non-targeting guides, as violin plots, highlighting the spread.

```{r}
plot_guide_violin(biscuit_fit)
```

`plot_gene_rank()` orders genes by their inferred effect size (`mu)` in descending order, visualizing the ranking and identifying the most significant hits.

```{r}
plot_gene_rank(biscuit_fit)
```

`plot_gene_volcano()` visualizes genes via their inferred effect size (`mu`) against their statistical significance.

```{r}
plot_gene_volcano(biscuit_fit)
```

`plot_phi_gamma_density()` shows posterior means of dispersion (`phi`) and how sample-level scaling (\`gamma\`) varies across samples.

```{r}
plot_phi_gamma_density(biscuit_fit)
```

`plot_mu_beta1` visualizes, for a given gene, the posterior distribution of the gene-level effect (`mu`) and the posterior means of guide-level effects (`beta1`) for all guides targeting that gene, showing how individual guides contribute to the overall gene effect.

```{r}
plot_mu_beta1(biscuit_fit, 'Zc3h12a')
```

## Conclusion

`biscuit` provides a robust workflow for analysis of CRISPR screens, combining Bayesian inference and intuitive visualizations to identify significant hits and quantify uncertainty at both the gene and guide levels.
